### PR TITLE
Allow the use of portable GCS settings and data

### PR DESCRIPTION
**Pull request by PTDreamer**
[Review on Reviewable](https://reviewable.io/reviews/d-ronin/dronin/268)

### CODE REVIEW

File: ground/gcs/src/libs/utils/pathutils.cpp [hunk 1]
```diff
@@ -33,13 +33,21 @@
 
 namespace Utils {
 
-QString PathUtils::settingsFilename = "";
+PathUtils * Utils::PathUtils::instance = 0;
```
Comment (mlyle) [inline ground/gcs/src/libs/utils/pathutils.cpp]: I think this should probably change to NULL.  0 used to be favored for pointers in C++, but with the emergence of C++11 and nullptr/nullptr_t NULL is safer in various ways.

File: ground/gcs/src/libs/utils/pathutils.cpp [hunk 2]
```diff
@@ -154,4 +162,35 @@ void PathUtils::setSettingsFilename(QString filename)
...
+void PathUtils::useGlobalSettings()
+{
+    setSettingsFilename(getGlobalSettingsFilePath());
+}
+
+bool PathUtils::useLocalSettings()
+{
+    bool result;
+    QString name = getLocalSettingsFilePath(result);
+    setSettingsFilename(name);
```
Comment (mlyle) [inline ground/gcs/src/libs/utils/pathutils.cpp:195]: Should we skip this if false, just to not surprise anyone doing stuff in the future with setSettingsFilename?

**Comment by PTDreamer** [comment]
Need to rewrite last commit comment. Will do it after review and force push.

File: ground/gcs/src/plugins/coreplugin/generalsettings.cpp [hunk 3]
```diff
@@ -161,12 +167,46 @@ void GeneralSettings::apply()
...
     m_autoConnect = m_page->checkAutoConnect->isChecked();
     m_autoSelect = m_page->checkAutoSelect->isChecked();
-    m_proxyType = m_page->proxyTypeCB->itemData(m_page->proxyTypeCB->currentIndex()).toInt();
+    m_proxyType = m_page->proxyTypeCB->currentData().toInt();
     m_proxyPort = m_page->portLE->text().toInt();
     m_proxyHostname = m_page->hostNameLE->text();
     m_proxyUser = m_page->userLE->text();
     m_proxyPassword = m_page->passwordLE->text();
     QNetworkProxy::setApplicationProxy (getNetworkProxy());
+    if(!m_usePortableSettings && m_page->cb_usePortableSettings->isChecked()) {
```
Comment (mlyle) [inline ground/gcs/src/plugins/coreplugin/generalsettings.cpp:180]: Just to make sure I understand-- to use portable settings, you need to start up once with global settings and then either import or change this?

File: ground/gcs/src/plugins/coreplugin/generalsettings.ui [hunk 4]
```diff
@@ -32,6 +32,16 @@
       <property name="verticalSpacing">
        <number>2</number>
       </property>
+      <item row="12" column="0">
```
Comment (mlyle) [inline ground/gcs/src/plugins/coreplugin/generalsettings.ui:35]: I haven't looked at this in QT creator yet but it looks like this one now occurs in the file twice? (Automatically connect a Tau Labs USB device).

File: ground/gcs/src/plugins/coreplugin/mainwindow.cpp [hunk 5]
```diff
@@ -331,31 +317,31 @@ void MainWindow::extensionsInitialized()
...
+        qs = settings;
         emit splashMessages(QString(tr("Loading default configuration from %1")).arg(filename));
         qDebug() << "Load default config from resource "<<filename;
     }
     qs->beginGroup("General");
+    if(usePortableSettings) {
+        bool writable;
+        Utils::PathUtils::getInstance()->getLocalSettingsFilePath(writable);
+        if(!writable) {
+            QMessageBox::warning(this, tr("Settings directory no writable"), tr("Reverting to global settings"));
```
Comment (mlyle) [inline ground/gcs/src/plugins/coreplugin/mainwindow.cpp]: It'd be nice if this was shared between the two places with similar code somehow.

Also I suggest "Settings directory is not writable"

**Comment by mlyle** [comment]
This looks good and is a good piece of functionality.  I'm not sure we want to take the risk of it for the first release, though.

**Comment by PTDreamer** [comment]
Review status: 0 of 23 files reviewed at latest revision, 4 unresolved discussions.

---

*[ground/gcs/src/libs/utils/pathutils.cpp, line 192 \[r1\]](https://reviewable.io:443/reviews/d-ronin/dronin/268#-K5aQ9VivwL2iyvx5HEt-r1-192:-K5a_srZHRfVJw5Va5P2:168568851) ([raw file](https://github.com/d-ronin/dronin/blob/27ac593616fb8066d0900700ec34fe90a2a3949a/ground/gcs/src/libs/utils/pathutils.cpp#L192)):*
Sound good

---

*[ground/gcs/src/plugins/coreplugin/generalsettings.cpp, line 176 \[r1\]](https://reviewable.io:443/reviews/d-ronin/dronin/268#-K5aQ9VivwL2iyvx5HEy-r1-176:-K5a_zz7qR9y12yZasuo:111582239) ([raw file](https://github.com/d-ronin/dronin/blob/27ac593616fb8066d0900700ec34fe90a2a3949a/ground/gcs/src/plugins/coreplugin/generalsettings.cpp#L176)):*
No. So the first time you start the gcs from clean settings it will ask you if you want to use global or portable settings. It will then write to that particular settings file creating it if necessary.
Upon further starts the presence of the portable settings file is checked and used if present.
Changing from portable to global settings renames the portable settings file (creating a backup) so it doesn't get detected and used on gcs start. It also checks if the global file exists and if it doesn't is creates one by copying the portable settings file. By doing this you avoid having the "no settings file found" dialog on next start.
Changing from global to portable does similarly the same on the opposed direction, with is what the piece of code we are commenting  does.

---

*[ground/gcs/src/plugins/coreplugin/generalsettings.ui, line 35 \[r1\]](https://reviewable.io:443/reviews/d-ronin/dronin/268#-K5aQ9VivwL2iyvx5HF--r1-35:-K5ad3ND4dzbNXwiJnmV:-1827381653) ([raw file](https://github.com/d-ronin/dronin/blob/27ac593616fb8066d0900700ec34fe90a2a3949a/ground/gcs/src/plugins/coreplugin/generalsettings.ui#L35)):*
Yes I see what you mean, seems like designer rearranged the file, if you look below the same lines were deleted, it happened with several other widgets. The ui seems ok though. 

---

*[ground/gcs/src/plugins/coreplugin/mainwindow.cpp, line 338 \[r1\]](https://reviewable.io:443/reviews/d-ronin/dronin/268#-K5aQ9VivwL2iyvx5HF0-r1-338:-K5adTQj0ie0qzEoXraR:160401764) ([raw file](https://github.com/d-ronin/dronin/blob/27ac593616fb8066d0900700ec34fe90a2a3949a/ground/gcs/src/plugins/coreplugin/mainwindow.cpp#L338)):*
ack

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/268)*

**Comment by PTDreamer** [comment]
Np, agreed

---

Review status: 0 of 23 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/268#-:-K5adc7YMb7GKLPgtLZv:441736275)*

**Comment by mlyle** [comment]
Review status: 0 of 23 files reviewed at latest revision, all discussions resolved.

---

*[ground/gcs/src/plugins/coreplugin/generalsettings.cpp, line 176 \[r1\]](https://reviewable.io:443/reviews/d-ronin/dronin/268#-K5aQ9VivwL2iyvx5HEy-r1-176:-K5agTmqB48Lx_tGCkle:-181202252) ([raw file](https://github.com/d-ronin/dronin/blob/27ac593616fb8066d0900700ec34fe90a2a3949a/ground/gcs/src/plugins/coreplugin/generalsettings.cpp#L176)):*
OK,  I'll try to try it out soon

---

*[ground/gcs/src/plugins/coreplugin/generalsettings.ui, line 35 \[r1\]](https://reviewable.io:443/reviews/d-ronin/dronin/268#-K5aQ9VivwL2iyvx5HF--r1-35:-K5agMazXF4Ef4vK4X75:-1840341977) ([raw file](https://github.com/d-ronin/dronin/blob/27ac593616fb8066d0900700ec34fe90a2a3949a/ground/gcs/src/plugins/coreplugin/generalsettings.ui#L35)):*
I was confused by automatically connect vs. automatically select.  my bad.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/268)*

**Comment by PTDreamer** [comment]
Applied review comments, rebased and force pushed.

**Comment by tracernz** [comment]
Reviewed 17 of 23 files at r1, 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

_[ground/gcs/src/libs/utils/pathutils.h, line 38 [r2]](https://reviewable.io:443/reviews/d-ronin/dronin/268#-K8XdgHts40s8yiYEn53:-K8XdgHts40s8yiYEn54:-202555831) ([raw file](https://github.com/d-ronin/dronin/blob/8f8d8daa82c5d64883d0dcb2fdc8ce7900244b2f/ground/gcs/src/libs/utils/pathutils.h#L38)):_
s/no/not/

---

_[ground/gcs/src/plugins/coreplugin/mainwindow.cpp, line 306 [r2]](https://reviewable.io:443/reviews/d-ronin/dronin/268#-K8Xf94B2KwqylOAwAwW:-K8Xf94B2KwqylOAwAwX:160412708) ([raw file](https://github.com/d-ronin/dronin/blob/af3165807536c335fc3d8310cd7e0d28126c70bc/ground/gcs/src/plugins/coreplugin/mainwindow.cpp#L306)):_
lol

---

_[ground/gcs/src/plugins/dial/dialgadgetconfiguration.cpp, line 63 [r2]](https://reviewable.io:443/reviews/d-ronin/dronin/268#-K8XfbbmHEaDss1z-nFc:-K8XfbbmHEaDss1z-nFd:-118535435) ([raw file](https://github.com/d-ronin/dronin/blob/8f8d8daa82c5d64883d0dcb2fdc8ce7900244b2f/ground/gcs/src/plugins/dial/dialgadgetconfiguration.cpp#L63)):_
No chance you want to clean these tabs up while you're here? :-P

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/268)_

**Comment by tracernz** [comment]
This looks good! I haven't tested yet.

**Comment by mlyle** [comment]
@PTDreamer Sorry that this has accumulated conflicts.. Glad to work on testing and getting merged once this is rebased.

**Comment by PTDreamer** [comment]
There was only one conflict on the settings ui. Fixed but not tested.

**Comment by mlyle** [comment]
@PTDreamer welcome back!

Is there any easy way to only ask the user about portable settings if an installer was not used?  e.g. only on the .zip and .tar?

**Comment by PTDreamer** [comment]
> @PTDreamer welcome back!
> 
> Is there any easy way to only ask the user about portable settings if an installer was not used? e.g. only on the .zip and .tar?

Thanks!
The latest commit makes it not show that option if the local settings directory is not writable, should do what you wanted right?

**Comment by mlyle** [comment]
Probably :)  I'll try it soon.

**Comment by mlyle** [comment]
Running from command line on Mac results in:

image

And a crash afterwards.. still studying

**Comment by mlyle** [comment]
Above error is actually form a regression introduced in #612

**Comment by mlyle** [comment]
Tried on OSX-- worked well both ways, except for this:

image

**Comment by tracernz** [comment]
Lets bring this forward and get it merged, been open way too long.

**Comment by mlyle** [comment]
Unfortunately, this adds complexity and isn't something people are clambering for.  Closing

**Pull request closed without merging**